wdl-format/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+//! A library for auto-formatting WDL code.
+
+#![warn(missing_docs)]
+#![warn(rust_2018_idioms)]
+#![warn(rust_2021_compatibility)]
+#![warn(missing_debug_implementations)]
+#![warn(clippy::missing_docs_in_private_items)]
+#![warn(rustdoc::broken_intra_doc_links)]
+
+use anyhow::Result;
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::Comment;
+use wdl_ast::Diagnostic;
+use wdl_ast::Direction;
+use wdl_ast::Document;
+use wdl_ast::Ident;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+use wdl_ast::SyntaxNode;
+use wdl_ast::Validator;
+use wdl_ast::Version;
+use wdl_ast::VersionStatement;
+
+mod comments;
+mod formatter;
+mod import;
+mod metadata;
+mod task;
+mod v1;
+mod workflow;
+
+use comments::format_inline_comment;
+use comments::format_preceding_comments;
+use formatter::Formatter;
+
+/// Newline constant used for formatting on windows platforms.
+#[cfg(windows)]
+pub const NEWLINE: &str = "\r\n";
+/// Newline constant used for formatting on non-windows platforms.
+#[cfg(not(windows))]
+pub const NEWLINE: &str = "\n";
+/// String terminator constant used for formatting.
+const STRING_TERMINATOR: char = '"';
+
+/// A trait for elements that can be formatted.
+pub trait Formattable {
+    /// Format the element and write it to the writer.
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        formatter: &mut Formatter,
+    ) -> std::fmt::Result;
+}
+
+impl Formattable for Version {
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        _formatter: &mut Formatter,
+    ) -> std::fmt::Result {
+        write!(writer, "{}", self.as_str())
+    }
+}
+
+impl Formattable for VersionStatement {
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        formatter: &mut Formatter,
+    ) -> std::fmt::Result {
+        let mut preamble_comments = Vec::new();
+        let mut lint_directives = Vec::new();
+        let comment_buffer = &mut String::new();
+        for sibling in self.syntax().siblings_with_tokens(Direction::Prev).skip(1) {
```

Comment:
Rowan includes `self` as the first sibling, so I'm skipping over `self`. Otherwise I'd need another arm in the `match` for `VersionStatementNode` that just ignores it.